feat : 상품 최신순 조회 기능 구현

yjjjwww opened the pull request:
### 변경사항
**AS-IS**
- 등록된 상품을 최신순으로 조회하는 기능 구현해습니다. 해당 기능은 Role에 상관없이 요청 가능합니다.
- 상품 이름, 가격, 설명, 수량, 이미지 정보, 카테고리 이름을 보여줍니다.
- 페이징 처리를 해서 page와 size를 함께 요청을 받으면 그에 맞는 결과를 가져옵니다.
- postman으로 API 테스트 완료했고, 테스트 코드로 올바른 응답이 나오는지 확인했습니다.

**TO-BE**
- 상품 가격순, 주문 횟수 순 조회하기

### 테스트
- [x] 테스트 코드
- [x] API 테스트

yjjjwww committed:
feat : [feature/product-search-latest] 상품 최신순으로 조회하기

- 페이징 처리를 하여 현재 페이지와 페이지 크기를 입력받아 그에 맞는 상품 정보를 최신순으로 조회
- 상품 이름, 상품 가격, 상품 설명, 상품 수량, 상품 이미지, 카테고리 이름 응답

yjjjwww committed:
test : [feature/product-search-latest] 상품 최신순으로 조회하기 테스트 코드 작성

- 원하는 size 만큼의 페이지 조회 결과를 가져오는지 확인
- Controller, Service 테스트 코드 작성

mingming-mentor commented on src/main/java/com/yjjjwww/yunmarket/product/controller/ProductController.java:
```diff
@@ -29,4 +33,12 @@ public ResponseEntity<String> registerProduct(
   }
+
+  @GetMapping("/search/latest")
+  public ResponseEntity<List<ProductInfo>> getLatestProducts(
```
상품 목록을 최신순으로 가지고 오는 기능으로 이해했어요. 요런 형태도 좋지만, paging을 사용하실거라면 sort 조건으로 createdDate를 넣으면 될듯해요. 참고하실수 있는 블로그 같이 첨부할게요 (https://itmining.tistory.com/139)
제가 말씀드린대로 구현이 된다면, 추후 다른 정렬 요구사항이 들어왔을때 보다 유연하게 대응할수 있을듯 합니다

yjjjwww replied:
@mingming-mentor
// ProuctRepository
public interface ProductRepository extends PagingAndSortingRepository<Product, Long> {

  List<Product> findAllBy(Pageable pageable);
}

// ProductServiceImpl
Pageable pageable = PageRequest.of(page - 1, size, Sort.by("createdDate").descending());

    List<Product> products = productRepository.findAllBy(pageable);

이렇게 하면 주문횟수순, 낮은 가격 순 조회를 할 때 새로운 메서드를 만들 필요가 없어 좋은 것 같습니다. 감사합니다.

mingming-mentor commented on src/main/java/com/yjjjwww/yunmarket/product/model/ProductInfo.java:
```diff
@@ -0,0 +1,20 @@
+@AllArgsConstructor
+public class ProductInfo {
+
+  String name;
```
private 붙여주세요

mingming-mentor commented on src/main/java/com/yjjjwww/yunmarket/product/service/ProductServiceImpl.java:
```diff
@@ -53,6 +58,28 @@ public void register(String token, ProductRegisterServiceForm form) {
+
+    List<ProductInfo> result = new ArrayList<>();
+
+    for (Product product : products) {
```
요렇게 작성하셔도 되고 stream을 이용하셔도 좋을 것 같아요. 그리고 entity -> dto 변환로직은 dto에 메소드로 만들어주시면 좋을듯해요. 추후 dto에 데이터가 추가될때 여러 메소드를 확인해보며 수정하지 않아도 dto에 정의해둔 메소드 수정만으로 대응이 가능할듯 합니당

mingming-mentor reviewed (approved):
수고 많으셨어요! 몇가지 리뷰드렸습니다. 이해가 잘 안가는 부분이 있다면 알려주세요

yjjjwww committed:
refactor : [feature/product-search-latest] 상품 최신순 조회 기능 코드 개선

- ProductRepository에서 상품 조회할 때 Sorting을 Pageable에 넘겨서 코드 재활용도를 높였습니다.
- ProductInfo 안에 private 추가
- Entity List -> DTO List 변환 로직을 ProductInfo에 메소드로 만들어줬습니다. stream을 이용했습니다.
